server/src/main/java/org/apache/druid/server/coordinator/helper/DruidCoordinatorRuleRunner.java

```diff
@@ -84,8 +85,10 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
     // find available segments which are not overshadowed by other segments in DB
     // only those would need to be loaded/dropped
     // anything overshadowed by served segments is dropped automatically by DruidCoordinatorCleanupOvershadowed
-    final Set<DataSegment> overshadowed = ImmutableDruidDataSource
-        .determineOvershadowedSegments(params.getAvailableSegments());
+    // If metadata store hasn't been polled yet, use empty overshadowed list
+    final Collection<DataSegment> overshadowed = Optional
+        .ofNullable(coordinator.getMetadataSegmentManager().findOvershadowedSegments())
```

Comment:
There are some concurrency issues: you may observe `overshadowedSegments` from a previous run. I think it would be easier to add `isOvershadowed` field to `DataSegment` directly in this PR because that's the plan anyway and that would allow avoiding concurrency issues.

Reply:
If it's acceptable to get updated overshadowedSegments in the next run of `DruidCoordinatorHelper#run`, then I think it might be ok. If not, then may be we should compute the overshadowed list here itself, like before, until the mutability of `DataSegment` is settled.

Reply:
Why are you reluctant to make changes to `DataSegment`? You refer to the design as "debated" or "not settled", well, we should resolve the debate and settle somewhere. I expressed my view in #7571. If you disagree with some parts of it, please respond in that issue.

I'm concerned with keeping the status quo because if `DataSegmentWithOvershadowedStatus` will leak into Druid 0.15 API, it will be much harder (transition PRs, compatibility, etc.) to remove it eventually. So I really want it to go away before Druid 0.15 is released.

Reply:
This is settled in [#7571(comment)](https://github.com/apache/incubator-druid/issues/7571#issuecomment-496681942). will be sending a flat format over the network for `SegmentWithOvershadowedStatus` without exposing the wrapping.